codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java

```diff
@@ -211,15 +233,94 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
             dstPort = header.readUnsignedShort();
         }
 
-        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort);
+        final List<HAProxyTLV> tlvs = readTlvs(header);
+
+        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort, tlvs);
+    }
+
+    private static List<HAProxyTLV> readTlvs(final ByteBuf header) {
+
+        //We need at least 4 byte for a valid TLV. If we have less bytes available, we need to ignore
+        if (header.readableBytes() < 4) {
+            return Collections.emptyList();
+        }
+
+        final List<HAProxyTLV> haProxyTLVs = new ArrayList<HAProxyTLV>();
+
+        while (header.isReadable()) {
+            final HAProxyTLV haProxyTLV = readNextTlv(header);
+
+            if (haProxyTLV == null) {
+                //We don't have a TLV available anymore, we can stop
+                break;
+            }
+            haProxyTLVs.add(haProxyTLV);
+            if (haProxyTLV instanceof HAProxySSLTLV) {
+                haProxyTLVs.addAll(((HAProxySSLTLV) haProxyTLV).getEncapsulatedTLVs());
+            }
+        }
+
+        return haProxyTLVs;
+    }
+
+    private static HAProxyTLV readNextTlv(final ByteBuf header) {
+
+        //We need at least 4 bytes for a TLV
+        if (header.readableBytes() < 4) {
+            return null;
+        }
+
+        final byte typeAsByte = header.readByte();
+        final HAProxyTLV.Type type = HAProxyTLV.Type.getType(typeAsByte);
+
+        final int length = header.readUnsignedShort();
+        switch (type) {
+
+        case PP2_TYPE_SSL:
+
+            final byte[] rawContent = copyByteBufToByteArray(header.duplicate(), length);
+            final ByteBuf byteBuf = header.readBytes(length);
+            final byte client = byteBuf.readByte();
+            final int verify = byteBuf.readInt();
+
+            final List<HAProxyTLV> encapsulatedTlvs = new ArrayList<HAProxyTLV>();
+            while (byteBuf.readableBytes() >= 4) {
+                final HAProxyTLV haProxyTLV = readNextTlv(byteBuf);
+                if (haProxyTLV == null) {
+                    break;
+                }
+                encapsulatedTlvs.add(haProxyTLV);
+            }
+
+            return new HAProxySSLTLV(verify, client, encapsulatedTlvs, rawContent);
+        //If we're not dealing with a SSL Type, we can use the same mechanism
+        case PP2_TYPE_ALPN:
+        case PP2_TYPE_AUTHORITY:
+        case PP2_TYPE_SSL_VERSION:
+        case PP2_TYPE_SSL_CN:
+        case PP2_TYPE_NETNS:
+        case OTHER:
+            final byte[] content = copyByteBufToByteArray(header, length);
+            return new HAProxyTLV(type, typeAsByte, content);
+        default:
+            return null;
+        }
+    }
+
+    private static byte[] copyByteBufToByteArray(final ByteBuf header, final int length) {
+        final byte[] content = new byte[length];
+        header.readBytes(content);
```

Comment:
use `header.getBytes(header.readerIndex, content);` so you not need to duplicate the `ByteBuf`